File: backend/app/src/main/kotlin/net/animeta/backend/serializer/PostSerializer.kt
package net.animeta.backend.serializer

import net.animeta.backend.dto.PostDTO
import net.animeta.backend.model.History
import org.springframework.stereotype.Service

@Service
class PostSerializer(val userSerializer: UserSerializer, val recordSerializer: RecordSerializer) {
    data class Options(val record: RecordSerializer.Options? = null, val user: UserSerializer.Options? = null)

    companion object {
        fun legacyOptions(includeRecord: Boolean = false, includeUser: Boolean = false) =
                Options(
                        record = if (includeRecord) RecordSerializer.legacyOptions() else null,
                        user = if (includeUser) UserSerializer.legacyOptions(includeCategories = false) else null
                )
    }

    fun serialize(history: History, options: Options): PostDTO {
        return PostDTO(
            id = history.id!!,
            record_id = history.record.id!!,
            status = history.status,
            status_type = history.status_type.name.toLowerCase(),
            comment = history.comment,
            updated_at = history.updatedAt?.toEpochMilli(),
            contains_spoiler = history.contains_spoiler,
            record = if (options.record != null) recordSerializer.serialize(history.record, options.record) else null,
            user = if (options.user != null) userSerializer.serialize(history.user, options = options.user) else null,
            rating = history.rating
        )
    }
}

File: backend-ts/src/serializers/post.serializer.ts
import { Injectable } from "@nestjs/common";
import { PostDTO } from "shared/types_generated";
import { History } from "src/entities/history.entity";
import { StatusType } from "src/entities/status_type";
import { RecordSerializer, RecordSerializerOptions } from "./record.serializer";
import { UserSerializer, UserSerializerOptions } from "./user.serializer";

export type PostSerializerOptions = {
  record?: RecordSerializerOptions;
  user?: UserSerializerOptions;
}

@Injectable()
export class PostSerializer {
  static legacyOptions({
    includeRecord = false,
    includeUser = false,
  }: {
    includeRecord?: boolean;
    includeUser?: boolean;
  } = {}): PostSerializerOptions {
    return {
      record: includeRecord ? RecordSerializer.legacyOptions() : undefined,
      user: includeUser ? UserSerializer.legacyOptions({ includeCategories: false }) : undefined,
    }
  }

  constructor(
    private recordSerializer: RecordSerializer,
    private userSerializer: UserSerializer,
  ) {
  }

  async serialize(history: History, options: PostSerializerOptions): Promise<PostDTO> {
    return {
      id: history.id,
      record_id: history.record.id,
      status: history.status,
      status_type: StatusType[history.status_type].toLowerCase(),
      comment: history.comment,
      updated_at: history.updated_at?.getTime() ?? null,
      contains_spoiler: history.contains_spoiler,
      record: options.record ? await this.recordSerializer.serialize(history.record, options.record ?? {}) : null,
      user: options.user ? await this.userSerializer.serialize(history.user, null, options.user ?? {}) : null,
      rating: history.rating,
    }
  }
}
